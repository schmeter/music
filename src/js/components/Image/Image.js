import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import autobind from 'autobind-decorator';

import configApp from '../../../config/app.json';

class Image extends React.Component {
    image = React.createRef();

    state = {
        loaded: false,
        error: false
    };

    @autobind
    handleLoad() {
        const { handleLoad } = this.props;
        this.setState({ loaded: true });
        handleLoad && handleLoad();
    }

    @autobind
    handleError() {
        const { error } = this.state;
        const image = this.image.current;
        this.setState({ error: true }, () => {
            if (!error) {
                image.src = configApp.fallbackImage;
            }
        });
    }

    render() {
        const { loaded } = this.state;
        const { className, src, alt } = this.props;
        return (
            <img
                className={classNames(
                    'image',
                    className,
                    { invisible: !loaded }
                )}
                src={src}
                alt={alt}
                ref={this.image}
                onLoad={this.handleLoad}
                onError={this.handleError}
            />
        );
    }
}

Image.propTypes = {
    src: PropTypes.string.isRequired,
    alt: PropTypes.string.isRequired,
    className: PropTypes.string,
    handleLoad: PropTypes.func
};

export default Image;
