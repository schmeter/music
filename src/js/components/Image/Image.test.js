import React from 'react';
import { mount } from 'enzyme';
import toJson from 'enzyme-to-json';

import Image from './Image';

describe('Image', () => {
  it('renders correctly', () => {
    const props = {
      src: 'test',
      alt: 'test',
      className: 'test',
      onLoad: jest.fn(),
    };
    const component = mount(<Image {...props} />);

    component.simulate('error');
    component.setState({ error: true });
    component.simulate('error');
    component.simulate('load');

    expect(toJson(component)).toMatchSnapshot();
    expect(props.onLoad).toHaveBeenCalled();
  });
});
