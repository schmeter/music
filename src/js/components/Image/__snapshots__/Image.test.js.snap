// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<Image
  alt="test"
  className="test"
  onLoad={
    [MockFunction] {
      "calls": Array [
        Array [],
      ],
      "results": Array [
        Object {
          "type": "return",
          "value": undefined,
        },
      ],
    }
  }
  src="test"
>
  <img
    alt="test"
    className="image test"
    onError={[Function]}
    onLoad={[Function]}
    src="test"
  />
</Image>
`;
